tensorflow/compiler/jit/mark_for_compilation_pass.cc

```diff
@@ -45,6 +45,42 @@ const char* const kXlaOutsideCompilationAttr = "_XlaOutsideCompilation";
 
 namespace {
 
+// An node will not be processed by XLA even it has the XLA kernel:
+// 1, In Tensorflow, conventionally, INT32 ops are regarded as shape or control ops, and
+// are registered on CPU only. XLA should follow the same rule, to avoid the potential
+// redundant memcopy in GPU_XLA case.
+// 2, Any OpKernel with more than 500 inputs is excluded. This is a temp workaround to avoid an
```

Comment:
I have no strong objection to this part.

That said it feels a bit like this is the wrong fix: whether or not the original TensorFlow operator has many inputs does not necessarily imply generated kernel has that many inputs, although I agree it is likely to lead to that for an operator such as a concatenation.

Perhaps give a concrete example as to a situation where this arises? What would need to happen before we could declare this issue fixed and remove the special case again?

Added Justin as a reviewer to comment on this part.

Reply:
> whether or not the original TensorFlow operator has many inputs does not necessarily imply 
> generated kernel has that many inputs, although I agree it is likely to lead to that for an 
> operator such as a concatenation.

Yes, that's exactly right, and why I'm also hesitant about fixing it in this way.

I think the right change is to fix the lowering in the XLA GPU backend to do something sane when it tries to construct a kernel with many inputs.  Essentially you'd want to add a new "many-args" calling convention.  A simple implementation of this calling convention would be:

  * Make the kernel take a single argument, namely `void** params`.  Then instead of referring to "paramN" in the kernel's body, we'd refer to params[N].
  * Before running the kernel, construct the params array in memory, probably as a H to D memcpy.

One might have to introduce a new `Thunk` subclass to accomplish this.  One would also have to teach buffer assignment that HLOs with the many-args calling convention need this additional buffer for their args.

#14507 is another bug report to the same effect.

Reply:
Justin,

As you and Hawkinsp metioned, fix 2 is actually a work-around and it could resolve our internal issue but not a fundamental solution. I agree we may solve it in a more principal way----add flexible enough support for op with many inputs althrough.

As to fix 1, May I know why "it is something of a historical accident that the TF classic GPU device doesn't have int32 kernels"? I am curious about this design situation. 
The reason for us to fix it is that for some complex models such as Bi-LSTM, without this fix, there will be many CPU2GPU data transfer which impede XLA performance.  If you have interest, I could explain it with more details.

Reply:
> A simple implementation of this calling convention would be:

Perhaps an even simpler implementation of this calling convention would be to have  void** of all of the root computation's params and outputs in GPU memory from the start.  That is, have one big array instead of one array per op with the large calling convention.

One would need to avoid creating this array if no HLOs have the large calling convention, to avoid that overhead.  But otherwise I expect it's simpler and probably more efficient if you have many ops with the large calling convention.